doghouse/server/storage.go

```diff
@@ -0,0 +1,115 @@
+package server
+
+import (
+	"fmt"
+
+	"golang.org/x/net/context"
+	"google.golang.org/appengine/datastore"
+)
+
+type Installation struct {
+	RepositoryFullName string // {owner}/{repo}
+	RepositoryID       int64
+	InstallationID     int
+	RepositoryToken    string
+}
+
+func newInstallationKey(ctx context.Context, repositoryName string) *datastore.Key {
+	kind := "Installation"
+	return datastore.NewKey(ctx, kind, repositoryName, 0, nil)
+}
+
+type CheckSuiteEvent struct {
+	Action     string `json:"action,omitempty"`
+	Repository struct {
+		ID       int64  `json:"id,omitempty"`
+		FullName string `json:"full_name,omitempty"`
+	} `json:"repository,omitempty"`
+	Installation struct {
+		ID int `json:"id,omitempty"`
+	} `json:"installation,omitempty"`
+}
+
+func SaveInstallationFromCheckSuite(ctx context.Context, c CheckSuiteEvent) error {
+	installation := Installation{
+		InstallationID:     c.Installation.ID,
+		RepositoryFullName: c.Repository.FullName,
+		RepositoryID:       c.Repository.ID,
+	}
+	return datastore.RunInTransaction(ctx, func(ctx context.Context) error {
+		ok, oldInst, err := getInstallation(ctx, installation.RepositoryFullName)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			// Not found. Generate repo token and save installation.
+			installation.RepositoryToken = GenerateRepositoryToken()
+			return saveInstallation(ctx, &installation)
+		}
+		// Found existing installation.
+		// Update InstallationID if and only if InstallationID is different.
+		if oldInst.InstallationID != installation.InstallationID {
+			oldInst.InstallationID = installation.InstallationID
+			return saveInstallation(ctx, oldInst)
+		}
+		return nil
+	}, nil)
+}
+
+func GetRepoToken(ctx context.Context, repoFullName string) (string, error) {
+	ok, inst, err := getInstallation(ctx, repoFullName)
+	if err != nil {
+		return "", err
+	}
+	if !ok {
+		return "", fmt.Errorf("repository token not found for %q", repoFullName)
+	}
+	return inst.RepositoryToken, nil
+}
+
+func GetOrUpdateRepoToken(ctx context.Context, repoFullName string, repoID int64, regenerate bool) (string, error) {
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported function GetOrUpdateRepoToken should have comment or be unexported